internal/overlord/servstate/manager.go

```diff
@@ -481,6 +485,44 @@ func (m *ServiceManager) CheckFailed(name string) {
 	}
 }
 
+// SetServiceArgs sets the service arguments provided by "pebble run --args"
+// to their respective services. It updates the service command in the top most
+// layer which contains the service in the plan.
+func (m *ServiceManager) SetServiceArgs(serviceArgs map[string][]string) error {
+	releasePlan, err := m.acquirePlan()
+	if err != nil {
+		return err
+	}
+	defer releasePlan()
+
+	layers := m.plan.Layers
+
+	for serviceName, args := range serviceArgs {
+		var layer *plan.Layer
+
+		// search for the topmost layer which contains the service
+		// assuming the layers are sorted by order ASC
+		for i := len(layers) - 1; i >= 0; i-- {
+			if _, ok := layers[i].Services[serviceName]; ok {
+				layer = layers[i]
+				break
+			}
+		}
+		if layer == nil {
+			return fmt.Errorf("service %q not found in plan", serviceName)
+		}
+
+		service := layer.Services[serviceName]
```

Comment:
Nitpick: To avoid extra `layer.Services` lookup, could be:
```
		var service *plan.Service

		// search for the topmost layer which contains the service
		// assuming the layers are sorted by order ASC
		for i := len(layers) - 1; i >= 0; i-- {
			if service = layers[i].Services[serviceName]; service != nil {
				break
			}
		}
		if service == nil {
			return fmt.Errorf("service %q not found in plan", serviceName)
		}
```

Reply:
Updated, thanks!